x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/scalar/ScalarFunction.java

```diff
@@ -110,31 +113,42 @@ protected ScriptTemplate scriptWithScalar(ScalarFunction scalar) {
     }
 
     protected ScriptTemplate scriptWithAggregate(AggregateFunction aggregate) {
-        String template = basicTemplate(aggregate);
-        return new ScriptTemplate(processScript(template),
-                paramsBuilder().agg(aggregate).build(),
-                dataType());
+        String template = "{}";
+        ParamsBuilder paramsBuilder = paramsBuilder().agg(aggregate);
+
+        DataType nullSafeCastDataType = null;
+        DataType dataType = aggregate.dataType();
+        if (dataType.name().equals("DATE") || dataType == DATETIME ||
+            // Aggregations on date_nanos are returned as string
+            aggregate.field().dataType() == DATETIME) {
+
+            template = "{sql}.asDateTime({})";
+        } else if (dataType.isInteger()) {
+            // MAX, MIN need to retain field's data type, so that possible operations on integral types (like division) work
+            // correctly -> perform a cast in the aggs filtering script, the bucket selector for HAVING.
+            // SQL function classes not available in QL: filter by name
+            String fn = aggregate.functionName();
+            if ("MAX".equals(fn) || "MIN".equals(fn)) {
+                nullSafeCastDataType = dataType;
+            } else if ("SUM".equals(fn)) {
+                // SUM(integral_type) requires returning a LONG value
+                nullSafeCastDataType = LONG;
+            }
+        }
+        if (nullSafeCastDataType != null) {
+            template = "{ql}.nullSafeCastNumeric({},{})";
+            paramsBuilder.variable(nullSafeCastDataType.name());
+        }
+        return new ScriptTemplate(processScript(template), paramsBuilder.build(), dataType());
     }
 
     // This method isn't actually used at the moment, since there is no grouping function (ie HISTOGRAM)
     // that currently results in a script being generated
     protected ScriptTemplate scriptWithGrouping(GroupingFunction grouping) {
-        String template = basicTemplate(grouping);
+        String template = "{}";
         return new ScriptTemplate(processScript(template),
-                paramsBuilder().grouping(grouping).build(),
-                dataType());
-    }
-
-    // FIXME: this needs to be refactored to account for different datatypes in different projects (ie DATE from SQL)
-    private String basicTemplate(Function function) {
```

Comment:
I've eventually removed this function since:
* the invocation from `scriptWithGrouping()` never exercise more than basic `{}` template initialisation (if even that, given the comment)
* most of its logic can be moved to `scriptWithAggregate()`.
* its FIXME seems to have been addressed?

Reply:
+1

The initial thinking was that by having one main method, further hooking/post-processing would be easier to add however so far that hasn't been the case and in this instance, proved too complicated and better served in the dedicated method.

Reply:
Update: I think the initial implementation was really just `return "{}"` - I think it's fine to have that per method.